yang/goldstone-nexthop.yang

```diff
@@ -0,0 +1,193 @@
+module goldstone-nexthop {
+
+  yang-version "1";
+
+  namespace "http://goldstone.net/yang/goldstone-nexthop";
+  prefix gs-nexthop;
+
+  import goldstone-interfaces { prefix gs-if; }
+
+  import ietf-inet-types {
+    prefix inet;
+  }
+
+  organization
+    "GoldStone";
+
+  description
+    "Goldstone Nexthop";
+
+  revision 2022-12-14 {
+    description
+      "Initial revision.";
+  }
+
+  typedef optional-ip-address {
+    type union {
+      type inet:ip-address;
+      type string {
+        length "0";
+      }
```

Comment:
Is this a standard way to represent an optional value in YANG?

Reply:
@ishidawataru 
Thank you for pointing it out.
Using `mandatory` statement may be more standard, so I will consider fixing it.

@ipi-claytonpascoal 
In my understanding, we need to specify IP address _**or**_ interface name for nexthop (gateway) when configuring static route on OcNOS.
Based on this, I propose to define `nexthop-config` group and `nexthop` list with `min-elements` as follows. What do you think?

```
  grouping nexthop-config {
    leaf nexthop-type {
      type nexthop-type;
      mandatory true;
      description
        "The nexthop type.";
    }

    leaf gateway {
      type inet:ipv4-address;
      description
        "The nexthop gateway address.";
    }

    leaf interface {
      type gs-if:interface-ref;
      description
        "The nexthop egress interface.";
    }

    leaf blackhole-type {
      type blackhole-type;
      description
        "A blackhole sub-type, if the nexthop is a blackhole type.";
    }

    leaf onlink {
      type boolean;
      default "false";
      description
        "Nexthop is directly connected.";
    }
  }

  ...

  grouping nexthop-top {
    container nexthops {
      description
        "GS nexthop object.";
      list nexthop {
        key "nexthop-type gateway interface";
        min-elements 1;
        description
          "A list of nexthop objects.";

        leaf nexthop-type {
          type leafref {
            path "../config/nexthop-type";
          }
        }

        leaf gateway {
          mandatory false;
          type leafref {
            path "../config/gateway";
          }
        }

        leaf interface {
          type leafref {
            path "../config/interface";
          }
        }

        container config {
          uses nexthop-config;
        }
        container state {
          config false;
          uses nexthop-config;
          uses nexthop-state;
        }
      }
    }
  }
```

Reply:
I am afraid that the `mandatory` statement does not help in this case. When the attributes are defined as keys of a list, they are automatically considered mandatory.
The challenge here is to support nexthop as a _gateway address_ or an _interface_ or, sometimes,  _gateway address + interface_.
I do not know how FRR handles these list keys.
OpenConfig uses an integer index as a key and has a union between IP address and string and another container for interface reference.
OcNOS has different route lists for each combination of nexthop keys.

In my opinion, the OpenConfig approach is more consistent.

Reply:
@ipi-claytonpascoal 
Thank you for your comments.
> I am afraid that the mandatory statement does not help in this case. When the attributes are defined as keys of a list, they are automatically considered mandatory.
> The challenge here is to support nexthop as a gateway address or an interface or, sometimes, gateway address + interface.

I see. Thank you for your advice, and it is helpful for me.
Umm, I try to consider and look into how OpenConfig and FRR handles the above..

Reply:
I agree to follow the OpenConfig approach.